# ================================================
# Titanic Data Analysis
# Tools: Python, Pandas, NumPy
# ================================================

In [53]:
import pandas as pd
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

In [54]:
# Always do these 5 things first!
print(df.head())


   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


In [55]:
print(df.shape)

(891, 12)


In [56]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


In [57]:
print(df.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


In [58]:
print(df.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


# ---- Step 1: Load Data ----
# ---- Step 2: Explore Data ----
# ---- Step 3: Clean Data ----
# ---- Step 4: Analysis ----
# ---- Step 5: Findings ----

In [59]:
# Q1: How many rows and columns?
print(df.shape)
# Q2: Which columns have missing values?
print(df.isnull().sum())


# Q3: What is the average age of passengers?
print(df["Age"].mean())
# Q4: How many passengers survived vs died?

#     Hint: df['Survived'].value_counts()
print(df['Survived'].value_counts())

# Q5: How many male vs female passengers?
#     Hint: df['Sex'].value_counts()
print(df['Sex'].value_counts())

(891, 12)
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64
29.69911764705882
Survived
0    549
1    342
Name: count, dtype: int64
Sex
male      577
female    314
Name: count, dtype: int64


In [60]:
# Step 1 — Fix Age (fill with median)
df['Age'] = df['Age'].fillna(df['Age'].median())

# Step 2 — Fix Embarked (fill with mode)
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])

# Step 3 — Drop Cabin (too many missing — 77%!)
df = df.drop(columns=['Cabin'])

# Step 4 — Verify no missing values left
print(df.isnull().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


In [61]:
# Q1: Survival rate by Gender
print(df.groupby('Sex')['Survived'].mean())

# Q2: Survival rate by Passenger Class
print(df.groupby('Pclass')['Survived'].mean())

# Q3: Average age by Passenger Class
print(df.groupby('Pclass')['Age'].mean())

# Q4: Average fare by Passenger Class
print(df.groupby('Pclass')['Fare'].mean())

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64
Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64
Pclass
1    36.812130
2    29.765380
3    25.932627
Name: Age, dtype: float64
Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64


In [79]:
# Q5: Survival rate by Embarked port

print(df.groupby('Embarked')['Survived'].mean())
print("====================================================")
# Q6: Average fare paid by male vs female
print(df.groupby('Sex')['Fare'].mean())
print("====================================================")
# Q7: How many passengers in each class?
print(df.groupby('Pclass')['Name'].count())
print("====================================================")
# Q8: Average age of survived vs died
print(df.groupby('Survived')['Age'].mean())
print("====================================================")
# Q9: Add AgeGroup column (Child/Adult/Senior)
df['AgeGroup']=df['Age'].apply(lambda x:'Child'if x<18 else ('Senior' if x>=40 else'Adult'))
print(df)
print("====================================================")
# Q10: Survival rate by AgeGroup
print(df.groupby('AgeGroup')["Survived"].mean())

Embarked
C    0.553571
Q    0.389610
S    0.339009
Name: Survived, dtype: float64
Sex
female    44.479818
male      25.523893
Name: Fare, dtype: float64
Pclass
1    216
2    184
3    491
Name: Name, dtype: int64
Survived
0    30.028233
1    28.291433
Name: Age, dtype: float64
     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                            

# ---- My Findings ----
# 1. Women had 74% survival rate vs men 19%
# 2. Class 1 passengers survived more (63%)
# 3. Children had highest survival (54%)
# 4. Port C had highest survival rate (55%)
# 5. Class 3 had most passengers (491)

Finding 1 — Port C had highest survival (55%)
Because most rich passengers boarded from
Cherbourg, France. Rich = Class 1 = more
access to lifeboats.

Finding 2 — Females paid more fare (44 vs 25)
Because females traveled in higher classes.
Higher class = more expensive tickets.

Finding 3 — Class 3 had most passengers (491)
Because most people were poor and couldn't
afford Class 1 or Class 2 tickets.

Finding 4 — Age didn't affect survival much
Died avg age = 30, Survived avg age = 28.
Only 2 years difference — so age was not
a deciding factor. Gender and Class mattered more!

Finding 5 — Children survived most (54%)
Because of "Women and Children First!" rule.
Children were given priority in lifeboats.